homeassistant/components/intellifire/config_flow.py

```diff
@@ -4,60 +4,138 @@
+                if self._discovered_hosts:
+                    # If there were discoverd hosts AND all got filtered out fire an error message
+                    errors["base"] = "already_discovered"
+                    LOGGER.debug("All discovered fireplaces have been configured")
 
-        try:
-            serial = await validate_input(self.hass, user_input[CONF_HOST])
-        except (ConnectionError, ClientConnectionError):
-            errors["base"] = "cannot_connect"
         else:
-            await self.async_set_unique_id(serial)
+            # There was User Input
+            placeholder = {CONF_HOST: user_input[CONF_HOST]}
+
             self._abort_if_unique_id_configured(
-                updates={CONF_HOST: user_input[CONF_HOST]}
+                updates={
+                    CONF_HOST: user_input[CONF_HOST],
+                }
             )
 
-            return self.async_create_entry(
-                title="Fireplace",
-                data={CONF_HOST: user_input[CONF_HOST]},
-            )
+            try:
+                # Validate the ip address
+                serial = await validate_host_input(user_input[CONF_HOST])
+            except (ConnectionError, ClientConnectionError):
+                errors["base"] = "cannot_connect"
+                local_schema = vol.Schema(
+                    {vol.Required(CONF_HOST, default=user_input[CONF_HOST]): str}
+                )
+            else:
+                await self.async_set_unique_id(serial)
+
+                return self.async_create_entry(
+                    title="Fireplace",
+                    data={
+                        CONF_HOST: user_input[CONF_HOST],
+                    },
+                )
+
+        return self.async_show_form(
+            step_id="local_config",
+            errors=errors,
+            description_placeholders=placeholder,
+            data_schema=local_schema,
+        )
+
+    async def async_step_pick_device(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Pick which device to configure."""
+
+        if user_input is not None:
+            # Update not_configured_hosts in order to trigger correct
+            # behavior in next step
+            self._not_configured_hosts = [user_input[CONF_HOST]]
+            return await self.async_step_local_config()
+
         return self.async_show_form(
-            step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
+            step_id="pick_device",
+            data_schema=vol.Schema(
+                {vol.Required(CONF_HOST): vol.In(self._discovered_hosts)}
+            ),
         )
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Start the user flow."""
+
+        # Launch fireplaces discovery
+        await self._find_fireplaces()
+
+        if len(self._not_configured_hosts) > 1:
+            return await self.async_step_pick_device()
```

Comment:
If the fireplace isn't discovered, there isn't a way to manually enter the ip address in this path.

Reply:
You are incorrect I'm afraid.

If nothing gets discovered (and I've tested this locally by forcing discovery to return an empty string)

This is the default value for these variables 

```python
        self._discovered_hosts: list[str] = []
        self._not_configured_hosts: list[str] = []
```

If no fireplaces are found both these variables will still be empty array

And so at this if block it will drop into:

```python
        return await self.async_step_local_config()
```

And because there is no `user_input` and no `_not_configured_hosts` and no `_discovered_hosts`

none of this logic will run



```python
   if user_input is None:

            if self._not_configured_hosts:
                user_input = {CONF_HOST: self._not_configured_hosts[0]}
                local_schema = vol.Schema(
                    {vol.Required(CONF_HOST, default=user_input[CONF_HOST]): str}
                )
            else:
                if self._discovered_hosts:
                    # If there were discoverd hosts AND all got filtered out fire an error message
                    errors["base"] = "already_discovered"
                    LOGGER.debug("All discovered fireplaces have been configured")
```

In which case it drop all the way through the user function to return 


```python
        return self.async_show_form(
            step_id="local_config",
            errors=errors,
            description_placeholders=placeholder,
            data_schema=local_schema,
        )
```


And to verify this for you visually -> I've commented out the following code block - to fully disable discovery. This case is also covered in  `async def test_no_discovery(...` test case.

<img width="498" alt="image" src="https://user-images.githubusercontent.com/6491743/158019302-54fbd3dc-8a88-490e-b432-55b52750ec21.png">

We have nothing discovered yet:

<img width="701" alt="image" src="https://user-images.githubusercontent.com/6491743/158019320-098cef1f-e1ab-4d26-9b67-30a28ee95ecb.png">

And as you can see it DOES prompt as-needed
<img width="585" alt="image" src="https://user-images.githubusercontent.com/6491743/158019362-35084cd9-4149-4cd5-b518-25822fe07ab4.png">

Reply:
I agree it will get you there, but it does so by passing input between steps which is what we want to avoid since it ends up with a more complex design

Reply:
Here is the scenario I'm concerned about

Fireplace 1= 1.2.3.4 (already configured)
Fireplace 2 =3.4.5.6
Fireplace 3 =4.5.6.7 (does not respond to the discovery request)

_discovered_hosts =  1.2.3.4, 3.4.5.6
_not_configured_hosts = 3.4.5.6

Since 4.5.6.7 is not seen by discovery there is no way to manually configure it